lib/id3c/api/datastore.py

```diff
@@ -259,35 +259,67 @@ def fetch_identifier_set(session: DatabaseSession, name: str) -> Any:
 def make_identifier_set(session: DatabaseSession, name: str, **fields) -> bool:
     """
     Create a new identifier set *name* in the backing database using *session*
-    if it doesn't already exist.
+    if it doesn't already exist, or update if it does exist.
+
+    If identifier set *name* does not exist, *use* keyword argument is required
+    and value will be set in the new record. If identifier set *name* exists,
+    *use* is optional and will be updated in the database if provided.  
 
     If *description* is provided as a keyword argument, its value is
-    set/updated in the database.
+    set/updated in the database. 
 
     Returns ``True`` if the set was created or updated and ``False`` if it
-    already existed.
+    already existed and was not updated.
+
+    Raises a :class:`BadRequestDatabaseError` exception if the database reports a
+    `constraint` error and a :class:`Forbidden` exception if the database reports a
+    `permission denied` error.
     """
     with session, session.cursor() as cursor:
-        # If I expected to have additional columns in the future, I'd build up
-        # the SQL dynamically, but I don't, so a simple conditional is enough.
-        #   -trs, 1 July 2021
-        if "description" in fields:
-            cursor.execute("""
-                insert into warehouse.identifier_set (name, description)
-                    values (%s, nullif(%s, ''))
-                    on conflict (name) do update
-                        set description = excluded.description
-                """, (name, fields["description"]))
-
-        else:
+        # *use* is required for insert; if *use* is not in fields then pull value
+        # from database record if it exists otherwise set to null
+        if "use" not in fields:
             cursor.execute("""
-                insert into warehouse.identifier_set (name)
-                    values (%s)
-                    on conflict (name) do nothing
+                select use from warehouse.identifier_set
+                where name = %s
                 """, (name,))
+            record = cursor.fetchone()
+            fields["use"] = record[0] if record else ''
+        
+        # build strings for generating SQL dynamically
+        columns_str = ", ".join(["name"] + list(fields.keys()))
+        values_str = "'" + "', '".join([name] + [fields[k] for k in fields.keys()]) + "'"
+        conflict_update_str = ", ".join([(k + " = excluded." + k) for k in fields.keys()])
+        conflict_where_str = " or ".join([("identifier_set." + k + " <> excluded." + k) for k in fields.keys()])
+
+        sql_str = "insert into warehouse.identifier_set ({}) \
+                    values ({}) \
+                    on conflict (name) do update set {} \
+                    where {}".format(columns_str, values_str, conflict_update_str, conflict_where_str)
+
+        sql_str = sql_str.replace("''","null")
+        try:
+            cursor.execute(sql_str)
```

Comment:
This introduces a SQL injection vulnerability by unsafely handling user input, specifically, the submitted field values.

The field names are risky, but technically ok for now, as our code is the sole provider of those. However, it's better practice for a function like this that accepts an arbitrary field list to treat it like untrusted input and handle accordingly. The reason is that the calling code is safe for now, but may change to not be safe in the future, and it won't be obvious at that point that a security vulnerability is being introduced.

Safe construction of a dynamic SQL statement can be done with the functions and classes provided by the [psycopg2.sql module](https://www.psycopg.org/docs/sql.html). There are several examples in ID3C and our other code bases.

Separately, I think this function might be a bit easier to read by introducing some conditional branching and using multiple formatted SQL statements (as it was before) instead of trying to fit everything into one.

Reply:
Thanks for pointing that out! I had switched to using `str.format(...)` in an attempt to generate a single SQL statement dynamically, but I can see why that's a bad idea. The link you provided has the information I needed.

I refactored again using a conditional statement with multiple SQL statements, but stuck with the existing pattern of hard coding SQL identifiers and passing query arguments, which should make it safe, more readable, and consistent with the other `datastore.py` functions.

The case where no `use` value is passed ended up requiring an additional select query to check if the record exists. If only the `description` is being updated on an existing record, switching to an update query made the most sense. The other challenge was keeping the function idempotent and returning a 204 code if nothing was changed. This required additional where clauses on `conflict (name) do update` and use of `coalesce` to compare null `description` values.